src/compiler/checker.ts

```diff
@@ -3993,6 +4027,37 @@ namespace ts {
             return links.resolvedType;
         }
 
+        // We do not perform supertype reduction on intersection types. Intersection types are created only by the &
+        // type operator and we can't reduce those because we want to support recursive intersection types. For example,
+        // a type alias of the form "type List<T> = T & { next: List<T> }" cannot be reduced during its declaration.

```

Comment:
So why do we do it for union types? Is it to cushion the loss of call/construct signatures?